src/host/allocations_cache.jl

```diff
@@ -0,0 +1,155 @@
+using Base.ScopedValues
+
+struct CacheAllocator{T <: AbstractGPUArray}
+    lock::ReentrantLock
+    busy::Dict{UInt64, Vector{T}} # hash((T, dims)) => GPUArray[]
+    free::Dict{UInt64, Vector{T}}
+end
+
+CacheAllocator(::Type{T}) where T = CacheAllocator(
+    ReentrantLock(),
+    Dict{UInt64, Vector{T}}(),
+    Dict{UInt64, Vector{T}}(),
+)
+
+function get_pool!(cache::CacheAllocator{T}, pool::Symbol, uid::UInt64) where T
+    pool = getproperty(cache, pool)
+    uid_pool = get(pool, uid, nothing)
+    if uid_pool ≡ nothing
+        uid_pool = Base.@lock cache.lock pool[uid] = T[]
+    end
+    return uid_pool
+end
+
+function alloc!(alloc_f, cache::CacheAllocator, ::Type{T}, dims::Dims{N}) where {T, N}
+    uid = hash((T, dims))
+    free_pool = get_pool!(cache, :free, uid)
+    busy_pool = get_pool!(cache, :busy, uid)
+
+    x = nothing
+
+    # No array available in `free` - call `alloc_f`.
+    isempty(free_pool) && (x = alloc_f())
+
+    # Otherwise, try fetching from `free`.
+    while !isempty(free_pool) && x ≡ nothing
+        tmp = pop!(free_pool)
+        # Array was manually freed via `unsafe_free!`.
+        storage(tmp).freed && continue
```

Comment:
Not necessarily only by calling `unsafe_free!`, but also by the GC, right?

Reply:
In this particular place only by `unsafe_free!`, because cached arrays are stored either in `free` or in `busy` arrays, thus preventing GC from collecting them.